src/hash-stream-validator.ts

```diff
@@ -54,12 +68,41 @@ class HashStreamValidator extends Transform {
     }
   }
 
-  _flush(callback: () => void) {
+  _flush(callback: (error?: Error | null | undefined) => void) {
     if (this.#md5Hash) {
       this.#md5Digest = this.#md5Hash.digest('base64');
     }
 
-    callback();
+    if (this.updateHashesOnly) {
+      callback();
+      return;
+    }
+
+    // If we're doing validation, assume the worst-- a data integrity
+    // mismatch. If not, these tests won't be performed, and we can assume
+    // the best.
+    // We must check if the server decompressed the data on serve because hash
+    // validation is not possible in this case.
+    let failed = this.crc32cEnabled || this.md5Enabled;
+
+    if (this.crc32cEnabled && this.crc32cExpected) {
+      failed = !this.test('crc32c', this.crc32cExpected);
+    }
+
+    if (this.md5Enabled && this.md5Expected) {
+      failed = !this.test('md5', this.md5Expected);
+    }
```

Comment:
I think I see the confusion - I think this block should only be checked in the presence of `this.crc32cExpected` or `this.md5Expected` (otherwise, if either aren't available then this would always fail).